Libraries/VtkVgCore/vtkVgSGIReader.cxx

```diff
@@ -0,0 +1,358 @@
+    }
+
+  // Open file
+  std::ifstream in{this->FileName, std::ifstream::in | std::ifstream::binary};
+  if (!in.good())
+    {
+    vtkErrorMacro("Failed to open input file.");
+    return 1;
+    }
+
+  // Read header
+  sgiImageHeader header;
+  in.read(reinterpret_cast<char*>(&header), sizeof(header));
+  if (!in.good())
+    {
+    vtkErrorMacro("Failed to read SGI header.");
+    return 1;
+    }
+  DiskToHardware(header.Magic);
+  DiskToHardware(header.StorageFormat);
+  DiskToHardware(header.BytesPerPixelChannel);
+  DiskToHardware(header.NumDimensions);
+  DiskToHardware(header.XDimension);
+  DiskToHardware(header.YDimension);
+  DiskToHardware(header.ZDimension);
+  DiskToHardware(header.MinValue);
+  DiskToHardware(header.MaxValue);
+  DiskToHardware(header.ColorMap);
+  if (header.Magic != 474)
+    {
+    vtkErrorMacro("Failed to read SGI file: bad magic.");
+    return 1;
+    }
+
+  // Check for supported formats
+#define NOT_SUPPORTED(pre, val, post) \
+    vtkErrorMacro( \
+      "Failed to read SGI file: " << pre << (pre && *pre ? " " : "") << val << \
+      post << (post && *post ? " " : "") << " is not supported."); \
+    return 1
+
+  if (header.BytesPerPixelChannel < 1 || header.BytesPerPixelChannel > 2)
+    {
+    NOT_SUPPORTED("", header.NumDimensions, "bytes per pixel channel");
+    }
+  if (header.StorageFormat != SGI_SF_Normal)
+    {
+    // TODO also support RLE?
+    NOT_SUPPORTED("storage format", header.StorageFormat, "");
+    }
+  if (header.ColorMap != SGI_CM_Normal)
+    {
+    // TODO also support dithered?
+    NOT_SUPPORTED("color map type", header.ColorMap, "");
+    }
+  if (header.NumDimensions < 2 || header.NumDimensions > 3)
+    {
+    NOT_SUPPORTED("dimensionality", header.NumDimensions, "");
+    }
+  if (header.NumDimensions == 2)
+    {
+    header.ZDimension = 1;
+    }
+
+  // Read image data
+  unsigned numImageBytes = header.XDimension * header.YDimension *
+                           header.ZDimension * header.BytesPerPixelChannel;
+  char* imageBytes = new char[numImageBytes];
+
+  in.read(imageBytes, numImageBytes);
+  if (!in.good())
+    {
+    vtkErrorMacro("Failed to read SGI file image data.");
+    return 1;
+    }
+
+  // Byte swap input data if needed
+  if (header.BytesPerPixelChannel == 2)
+    {
+    for (unsigned i = 0; i < numImageBytes; ++i)
+      {
+      DiskToHardware(*reinterpret_cast<uint16_t*>(imageBytes + i));
+      }
+    }
+
+  // Convert to vtkImageData
+  const int xExtent = header.XDimension - 1;
+  const int yExtent = header.YDimension - 1;
+  const ptrdiff_t planeStride = header.XDimension * header.YDimension *
+                                header.BytesPerPixelChannel;
+  const int pdt = (header.BytesPerPixelChannel == 1 ? VTK_UNSIGNED_CHAR
+                                                    : VTK_UNSIGNED_SHORT);
+
+  auto components = vtkSmartPointer<vtkImageAppendComponents>::New();
+  std::vector<vtkSmartPointer<vtkImageImport>> importers;
```

Comment:
The contents of this vector never get used... is it just here to hold a smart pointer reference while `components->Update()` is run?